File: src/app/services/catalogos.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from './api.service';

export interface CatSector {
  sectorCategoryId: number;
  sector: string;
}

export interface CatSubsector {
  subsectorCategoryId: number;
  subsector: string;
  sectorCategoryId?: number;
  status: number;
  sectorCategory?: CatSector;
}

export interface CatPrioridad {
  priorityCategoryId: number;
  priority: string;
}

export interface CatCondicion {
  conditionCategoryId: number;
  condition: string;
}

export interface CatInformacion {
  informationCategoryId: number;
  information: string;
}

export interface CatMunicipio {
  municipalityCategoryId: number;
  municipality: string;
  delegationId?: number;
}

export interface CatDelegacion {
  delegationCategoryId: number;
  delegation: string;
}

export interface CatMedidaSeguridad {
  measureCategoryId: number;
  measure: string;
  status: number;
}

@Injectable({
  providedIn: 'root',
})
export class CatalogosService {
  private readonly apiUrl = 'catalogos';

  constructor(private api: ApiService) {}

  async getSectors(): Promise<CatSector[]> {
    try {
      const response = await this.api.fetch(`${this.apiUrl}/sectores`);
      if (!response.ok) {
        throw new Error(`Error al get sectors: ${response.statusText}`);
      }
      return await response.json();
    } catch (error) {
      return [];
    }
  }

  async getSubsectors(): Promise<CatSubsector[]> {
    try {
      const response = await this.api.fetch(`${this.apiUrl}/subsectores`);
      if (!response.ok) {
        throw new Error(`Error al get subsectors: ${response.statusText}`);
      }
      return await response.json();
    } catch (error) {
      return [];
    }
  }

  async getSubsectorsBySector(sectorId: number): Promise<CatSubsector[]> {
    try {
      const response = await this.api.fetch(
        `${this.apiUrl}/subsectores/sector/${sectorId}`,
      );
      if (!response.ok) {
        throw new Error(`Error al get subsectors: ${response.statusText}`);
      }
      return await response.json();
    } catch (error) {
      return [];
    }
  }

  async getPriorities(): Promise<CatPrioridad[]> {
    try {
      const response = await this.api.fetch(`${this.apiUrl}/prioridades`);
      if (!response.ok) {
        throw new Error(`Error al get priorities: ${response.statusText}`);
      }
      return await response.json();
    } catch (error) {
      return [];
    }
  }

  async getConditions(): Promise<CatCondicion[]> {
    try {
      const response = await this.api.fetch(`${this.apiUrl}/condiciones`);
      if (!response.ok) {
        throw new Error(`Error al get conditions: ${response.statusText}`);
      }
      return await response.json();
    } catch (error) {
      return [];
    }
  }

  async getInformationItems(): Promise<CatInformacion[]> {
    try {
      const response = await this.api.fetch(`${this.apiUrl}/informaciones`);
      if (!response.ok) {
        throw new Error(
          `Error al get informationItems: ${response.statusText}`,
        );
      }
      return await response.json();
    } catch (error) {
      return [];
    }
  }

  async getMunicipalities(): Promise<CatMunicipio[]> {
    try {
      const response = await this.api.fetch(`${this.apiUrl}/municipios`);
      if (!response.ok) {
        throw new Error(`Error al get municipalities: ${response.statusText}`);
      }
      return await response.json();
    } catch (error) {
      return [];
    }
  }

  async getDelegations(): Promise<CatDelegacion[]> {
    try {
      const response = await this.api.fetch(`${this.apiUrl}/delegaciones`);
      if (!response.ok) {
        throw new Error(
          `Error al get delegations: ${response.statusText}`,
        );
      }
      return await response.json();
    } catch (error) {
      return [];
    }
  }

  async getSecurityMeasures(): Promise<CatMedidaSeguridad[]> {
    try {
      const response = await this.api.fetch(`${this.apiUrl}/medidas-seguridad`);

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(
          `Error al get measures de seguridad: ${response.statusText}`,
        );
      }

      const data = await response.json();
      return data;
    } catch (error) {
      return [];
    }
  }
}

File: backend/Services/CatalogosService.cs
using IRS.API.Data;
using IRS.API.Interfaces;
using IRS.API.Models;
using Microsoft.EntityFrameworkCore;

namespace IRS.API.Services;

public class CatalogsService : ICatalogosService
{
    private readonly IRSDbContext _context;

    public CatalogsService(IRSDbContext context)
    {
        _context = context;
    }

    public async Task<List<SectorCategory>> GetSectorsAsync()
    {
        return await _context.Sectors
            .OrderBy(s => s.Sector)
            .ToListAsync();
    }

    public async Task<List<CatSubsector>> GetSubsectorsAsync()
    {
        return await _context.Subsectors
            .Include(ss => ss.SectorCategory)
            .Where(ss => ss.Status == 1)
            .OrderBy(ss => ss.Subsector)
            .ToListAsync();
    }

    public async Task<List<CatSubsector>> GetSubsectorsBySectorAsync(int sectorId)
    {
        return await _context.Subsectors
            .Include(ss => ss.SectorCategory)
            .Where(ss => ss.SectorCategoryId == sectorId && ss.Status == 1)
            .OrderBy(ss => ss.Subsector)
            .ToListAsync();
    }

    public async Task<List<CatPrioridad>> GetPrioritiesAsync()
    {
        return await _context.Priorities
            .OrderBy(p => p.Priority)
            .ToListAsync();
    }

    public async Task<List<CatCondicion>> GetConditionsAsync()
    {
        return await _context.Conditions
            .OrderBy(c => c.Condition)
            .ToListAsync();
    }

    public async Task<List<CatInformacion>> GetInformationItemsAsync()
    {
        return await _context.InformationItems
            .OrderBy(i => i.Information)
            .ToListAsync();
    }

    public async Task<List<CatMunicipio>> GetMunicipalitiesAsync()
    {
        return await _context.Municipalities
            .OrderBy(m => m.Municipality)
            .ToListAsync();
    }

    public async Task<List<CatDelegacion>> GetDelegationsAsync()
    {
        return await _context.Delegations
            .OrderBy(d => d.Delegation)
            .ToListAsync();
    }

    public async Task<List<CatMedidaSeguridad>> GetSecurityMeasuresAsync()
    {
        return await _context.SecurityMeasures
            .Where(m => m.Status == 1)
            .OrderBy(m => m.Measure)
            .ToListAsync();
    }
}
